catalog/resource_share.go

```diff
@@ -0,0 +1,172 @@
+}
+
+type ShareDataChange struct {
+	Action     string           `json:"action"`
+	DataObject SharedDataObject `json:"data_object"`
+}
+
+type ShareUpdates struct {
+	Updates []ShareDataChange `json:"updates"`
+}
+
+type Shares struct {
+	Shares []ShareInfo `json:"shares"`
+}
+
+func (a SharesAPI) list() (shares Shares, err error) {
+	err = a.client.Get(a.context, "/unity-catalog/shares", nil, &shares)
+	return
+}
+
+func (a SharesAPI) create(ci *ShareInfo) error {
+	return a.client.Post(a.context, "/unity-catalog/shares", ci, ci)
+}
+
+func (a SharesAPI) get(name string) (ci ShareInfo, err error) {
+	err = a.client.Get(a.context, "/unity-catalog/shares/"+name+"?include_shared_data=true", nil, &ci)
+	return
+}
+
+func (a SharesAPI) update(name string, su ShareUpdates) error {
+	if len(su.Updates) == 0 {
+		return nil
+	}
+	return a.client.Patch(a.context, "/unity-catalog/shares/"+name, su)
+}
+
+func (a SharesAPI) delete(name string) error {
+	return a.client.Delete(a.context, "/unity-catalog/shares/"+name, nil)
+}
+
+func shareObjectsToShareChange(si ShareInfo, action string) ShareUpdates {
+	var changes []ShareDataChange
+	for _, obj := range si.Objects {
+		changes = append(changes, ShareDataChange{
+			Action:     action,
+			DataObject: obj,
+		})
+	}
+	return ShareUpdates{
+		Updates: changes,
+	}
+}
+
+func getResourceShareMap(si ShareInfo) map[string]SharedDataObject {
+	m := make(map[string]SharedDataObject, len(si.Objects))
+	for _, sdo := range si.Objects {
+		m[sdo.Name] = sdo
+	}
+	return m
+}
+
+func (si ShareInfo) Diff(other ShareInfo) []ShareDataChange {
+	beforeMap := getResourceShareMap(si)
+	afterMap := getResourceShareMap(other)
+	changes := []ShareDataChange{}
+	// not in after so remove
+	for _, beforeSdo := range si.Objects {
+		_, exists := afterMap[beforeSdo.Name]
+		if exists {
+			continue
+		}
+		changes = append(changes, ShareDataChange{
+			Action:     ShareRemove,
+			DataObject: beforeSdo,
+		})
+	}
+
+	// not in before so add
+	for _, afterSdo := range other.Objects {
+		_, exists := beforeMap[afterSdo.Name]
+		if exists {
+			continue
+		}
+		changes = append(changes, ShareDataChange{
+			Action:     ShareAdd,
+			DataObject: afterSdo,
+		})
+	}
+	return changes
+}
+
+func ResourceShare() *schema.Resource {
+	shareSchema := common.StructToSchema(ShareInfo{}, func(m map[string]*schema.Schema) map[string]*schema.Schema {
+		return m
+	})
+	return common.Resource{
+		Schema: shareSchema,
+		Create: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			var si ShareInfo
+			common.DataToStructPointer(d, shareSchema, &si)
+			if err := NewSharesAPI(ctx, c).create(&si); err != nil {
+				return err
+			}
+			//can only create empty share, objects have to be added using update API
+			shareChanges := shareObjectsToShareChange(si, ShareAdd)
+			if err := NewSharesAPI(ctx, c).update(si.Name, shareChanges); err != nil {
+				return err
```

Comment:
this error will cause the inability to create a share with the same name after.

please cleanup the share you've created on line 138 if this error happens.